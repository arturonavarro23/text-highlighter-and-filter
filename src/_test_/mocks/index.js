
export const defaultState = {
  text: 'Test words',
  colors: ['red', 'yellow', 'green'],
  selectedColor: 'red',
  selectedFilters: ['red', 'yellow'],
  ranges: []
};
