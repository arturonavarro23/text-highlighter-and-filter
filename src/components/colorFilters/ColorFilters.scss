@import '../../scss/abstracts/variables';

.color-filters {
  h4 {
    margin: 10px 0;
    text-transform: capitalize;
  }

  span {
    display: inline-block;
  }

  .green {
    background-color: $green-color;
  }

  .yellow {
    background-color: $yellow-color;
  }

  .red {
    background-color: $red-color;
  }
}
