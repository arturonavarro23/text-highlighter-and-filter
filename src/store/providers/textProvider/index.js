import React, { createContext, useContext, useReducer } from 'react';

export const TextContext = createContext();

export const TextProvider = ({ reducer, initialState, children }) => (
  <TextContext.Provider value={useReducer(reducer, initialState)}>
    {children}
  </TextContext.Provider>
);

export const useStateValue = () => useContext(TextContext);
