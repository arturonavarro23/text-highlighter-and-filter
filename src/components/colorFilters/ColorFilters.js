import React from 'react';
import { useStateValue } from '../../store/providers/textProvider';
import { removeFilter, addFilter } from '../../store/actions';
import Button from '../common/button';

import './ColorFilters.scss';

const ColorFilters  = () => {
  const [{
    colors,
    selectedFilters,
    ranges,
    text
  }, dispatch] = useStateValue();

  const onClick = color => () => {
    // check if filter is already active to toggle the state
    if (selectedFilters.includes(color)) {
      dispatch(removeFilter(color));
    } else {
      dispatch(addFilter(color));
    }
  };

  const renderButton = color => (
    <Button
      key={color}
      variant={color}
      onClick={onClick(color)}
      selected={selectedFilters.includes(color)}
    />
  );

  const printFilteredText = color => {
    // get words from active color filter
    const highlights = ranges.filter(r => r.color === color)
      .map(t => {
        const word = text.substring(t.start, t.end);
        return(
          <React.Fragment key={`${t.start}-${t.end}`}>
            <span className={color}>
              {word}
            </span>
            <br />
          </React.Fragment>
        );
      });

    return (
      <React.Fragment key={`${color}-text`}>
        <h4>
          {color}
        </h4>
        {highlights}
      </React.Fragment>
    );
  };

  return(
    <div className="color-filters">
      <h1>
        Filters
      </h1>
      {colors.map(renderButton)}
      <div className="filtered-text">
        {selectedFilters.map(printFilteredText)}
      </div>
    </div>
  );
};

export default ColorFilters;
