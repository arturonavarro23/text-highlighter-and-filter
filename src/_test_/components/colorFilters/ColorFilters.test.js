import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import * as TextProvider from '../../../store/providers/textProvider';
import ColorFilters from '../../../components/colorFilters';
import { defaultState } from '../../mocks';

describe('<colorFilters> test', () => {
  it('Snapshot should render correctly', () => {
    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState]);

    const colorSelectorTree = renderer
      .create(<ColorFilters />)
      .toJSON();

    expect(colorSelectorTree).toMatchSnapshot();
  });
  it('Should render all buttons', () => {
    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState]);

    const colorSelector = mount(<ColorFilters />);
    expect(colorSelector.find('button')).toHaveLength(defaultState.colors.length);
  });
  it('Should have selected button', () => {
    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState]);

    const colorSelector = mount(<ColorFilters />);
    expect(colorSelector.find('.selected')).toHaveLength(defaultState.selectedFilters.length);
  });
  it('onClick should be called', () => {

    const mockFn = jest.fn().mockName('dispatch');

    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [defaultState, mockFn]);

    const colorSelector = mount(<ColorFilters />);
    const selectedButton = colorSelector.find('.selected').last();
    selectedButton.simulate("click");
    expect(mockFn).toHaveBeenCalled();
  });
  it('Should render highlights', () => {

    const newState = Object.assign({}, defaultState, {
      ranges: [{ start: 0, end: 10, color: 'red'}],
      selectedFilters: ['red']
    });

    jest
      .spyOn(TextProvider, 'useStateValue')
      .mockImplementation(() => [newState]);

    const colorSelector = mount(<ColorFilters />);
    const highlight = colorSelector.find('.red');
    expect(highlight.html()).toBe('<span class="red">Test words</span>');
  });
});
