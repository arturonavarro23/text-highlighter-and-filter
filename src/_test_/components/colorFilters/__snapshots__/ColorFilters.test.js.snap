// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<colorFilters> test Snapshot should renders correctly 1`] = `
<div
  className="color-filters"
>
  <h1>
    Filters
  </h1>
  <button
    className="btn btn--red selected"
    onClick={[Function]}
  />
  <button
    className="btn btn--yellow selected"
    onClick={[Function]}
  />
  <button
    className="btn btn--green"
    onClick={[Function]}
  />
  <div
    className="filtered-text"
  >
    <h4>
      red
    </h4>
    <h4>
      yellow
    </h4>
  </div>
</div>
`;
